pkg/controllers/node/termination/terminator/eviction.go

```diff
@@ -71,117 +75,107 @@ func IsNodeDrainError(err error) bool {
+			MaxConcurrentReconciles: 100,
+		}).
+		Complete(reconcile.AsReconciler(m.GetClient(), q))
 }
 
 // Add adds pods to the Queue
-func (q *Queue) Add(node *corev1.Node, pods ...*corev1.Pod) {
-	q.mu.Lock()
-	defer q.mu.Unlock()
+func (q *Queue) Add(pods ...*corev1.Pod) {
+	q.Lock()
+	defer q.Unlock()
 
 	for _, pod := range pods {
-		qk := NewQueueKey(pod, node.Spec.ProviderID)
+		qk := NewQueueKey(pod)
 		if !q.set.Has(qk) {
 			q.set.Insert(qk)
-			q.TypedRateLimitingInterface.Add(qk)
+			q.source <- event.TypedGenericEvent[*corev1.Pod]{Object: pod}
 		}
 	}
 }
 
-func (q *Queue) Has(node *corev1.Node, pod *corev1.Pod) bool {
-	q.mu.Lock()
-	defer q.mu.Unlock()
+func (q *Queue) Has(pod *corev1.Pod) bool {
+	q.Lock()
+	defer q.Unlock()
 
-	return q.set.Has(NewQueueKey(pod, node.Spec.ProviderID))
+	return q.set.Has(NewQueueKey(pod))
 }
 
-func (q *Queue) Reconcile(ctx context.Context) (reconcile.Result, error) {
-	ctx = injection.WithControllerName(ctx, "eviction-queue")
-	// Check if the queue is empty. client-go recommends not using this function to gate the subsequent
-	// get call, but since we're popping items off the queue synchronously, there should be no synchonization
-	// issues.
-	if q.TypedRateLimitingInterface.Len() == 0 {
-		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
-	}
-	// Get pod from queue. This waits until queue is non-empty.
-	item, shutdown := q.TypedRateLimitingInterface.Get()
-	if shutdown {
-		return reconcile.Result{}, fmt.Errorf("EvictionQueue is broken and has shutdown")
-	}
-
-	defer q.TypedRateLimitingInterface.Done(item)
+func (q *Queue) Reconcile(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, q.Name())
 
+	if !q.Has(pod) {
+		//This is a different pod than in the queue, we should exit without evicting
+		return reconcile.Result{}, nil
+	}
 	// Evict the pod
-	if q.Evict(ctx, item) {
-		q.TypedRateLimitingInterface.Forget(item)
-		q.mu.Lock()
-		q.set.Delete(item)
-		q.mu.Unlock()
-		return reconcile.Result{RequeueAfter: singleton.RequeueImmediately}, nil
+	if err := q.Evict(ctx, pod); err != nil {
+		return reconcile.Result{}, err
 	}
 
-	// Requeue pod if eviction failed
-	q.TypedRateLimitingInterface.AddRateLimited(item)
-	return reconcile.Result{RequeueAfter: singleton.RequeueImmediately}, nil
+	q.Lock()
+	defer q.Unlock()
+	q.set.Delete(NewQueueKey(pod))
+	return reconcile.Result{}, nil
 }
 
-// Evict returns true if successful eviction call, and false if there was an eviction-related error
-func (q *Queue) Evict(ctx context.Context, key QueueKey) bool {
-	ctx = log.IntoContext(ctx, log.FromContext(ctx).WithValues("Pod", klog.KRef(key.Namespace, key.Name)))
+// Evict returns nil if successful eviction call, and an error if there was an eviction-related error
+func (q *Queue) Evict(ctx context.Context, pod *corev1.Pod) error {
+	ctx = log.IntoContext(ctx, log.FromContext(ctx).WithValues("Pod", klog.KRef(pod.Namespace, pod.Name)))
```

Comment:
This should automatically be injected by the Reconcile() method so there's actually no need to add it to the context here